src/context.jl

```diff
@@ -1,105 +1,206 @@
     return findlast(x -> !ismissing(ctx, x), data)
 end
 
 """
-    findnext(ctx::Context, data::AbstractVector) -> Int
+    findnext(ctx::AbstractContext, data::AbstractVector) -> Int
 
 Returns the next not missing index in `data`.
 
 # Arguments
-* `ctx::Context`: the context to pass into `ismissing`
+* `ctx::AbstractContext`: the context to pass into `ismissing`
 * `data::AbstractVector`: the data array to search
 
 # Returns
 * `Int`: the next index in `data` that isn't missing
 """
-function Base.findnext(ctx::Context, data::AbstractVector, idx::Int)
+function Base.findnext(ctx::AbstractContext, data::AbstractVector, idx::Int)
     return findnext(x -> !ismissing(ctx, x), data, idx)
 end
+
+"""
+    Context
+
+Records base information about the missing data and assume all observations are equally
+weighted.
+
+# Fields
+* `n::Int`: number of observations
+* `count::Int`: number of missing values found
+* `limit::Float64`: allowable limit for missing values to impute
+* `is_missing::Function`: returns a Bool if the value counts as missing
+* `on_complete::Function`: a function to run when imputation is complete
+"""
+mutable struct Context <: AbstractContext
+    num::Int
+    count::Int
+    limit::Float64
+    is_missing::Function
+    on_complete::Function
+end
+
+function Context(;
+    limit::Float64=0.1,
+    is_missing::Function=ismissing,
+    on_complete::Function=complete
+)
+    Context(0, 0, limit, is_missing, on_complete)
+end
+
+function (ctx::Context)(f::Function)
+    _ctx = copy(ctx)
+    _ctx.num = 0
+    _ctx.count = 0
+
+    result = f(_ctx)
+    ctx.on_complete(_ctx)
+    return result
+end
+
+Base.copy(x::Context) = Context(x.num, x.count, x.limit, x.is_missing, x.on_complete)
+
+function missing_update!(ctx::Context, miss)
+    ctx.num += 1
+
+    if miss
+        ctx.count += 1
+    end
+end
+
+function complete(ctx::Context)
+    missing_ratio = ctx.count / ctx.num
+
+    if missing_ratio > ctx.limit
+        throw(ImputeError(
+            "More than $(ctx.limit * 100)% of values were missing ($missing_ratio)."
+        ))
+    end
+end
+
+
+"""
+    WeightedContext
+
+Records information about the missing data relative to a set of weights.
+This context type can be useful if some missing observation are more important than others
+(e.g., more recent observations in time series datasets)
+
+# Fields
+* `num::Int`: number of observations
+* `s::Float64`: sum of missing values weights
+* `limit::Float64`: allowable limit for missing values to impute
+* `is_missing::Function`: returns a Bool if the value counts as missing
+* `on_complete::Function`: a function to run when imputation is complete
+* `wv::AbstractWeights`: a set of statistical weights to use when evaluating the importance
+  of each observation
+"""
+mutable struct WeightedContext <: AbstractContext
+    num::Int
+    s::Float64
+    limit::Float64
+    is_missing::Function
+    on_complete::Function
```

Comment:
We're actively using the `is_missing` in the existing tests. I'm less set on `on_complete`. My initial thought was that folks would want to switch the `ImputeError` to a warning, but I think just catching the exception is probably fine. Also, most more sophisticated behaviour would require a custom type anyways (e.g., progressive warning system with multiple thresholds).

Reply:
Hmm, yes I think the argument for `on_complete` being not configuable is strong
For now we can leave `ismissing` configuable.
but we might want to think about requring preprocessing later.
(e.g. make the user for normalize `NaN`s and `nothing`s and `""` and `-999` etc. via a LazyMapArray or generator or something).

But let's save that discussion for an issue.

> Also, most more sophisticated behaviour would require a custom type anyways (e.g., progressive warning system with multiple thresholds).

I do wonder if we can enrich the `ImputeError` type to have this info some how.
But that is another PR

Reply:
I think the `on_complete` function would be more helpful if we took both the data and the context before returning. This would give people the option of changing what is returned based on their checks without needing to catch exceptions.

Reply:
So our main execution would look something like:

```
function (ctx::Context)(f::Function)
    _ctx = empty(ctx)
    return ctx.on_complete(f(_ctx), _ctx)
end
```

Reply:
I'm going to try it like this for a week or so. If we decide that it doesn't provide any advantage then I'll remove it before hitting 1.0.